<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.475">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Aki Vehtari et al.">

<title>Bayesian Data Analysis - Notebook for Assignment 6</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title">Notebook for Assignment 6</h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Bayesian Data Analysis</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">General instructions</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment_instructions.html" class="sidebar-item-text sidebar-link">Assignment instructions</a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">Assignments</a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment1.html" class="sidebar-item-text sidebar-link">Assignment 1</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment2.html" class="sidebar-item-text sidebar-link">Assignment 2</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment3.html" class="sidebar-item-text sidebar-link">Assignment 3</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment4.html" class="sidebar-item-text sidebar-link">Assignment 4</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment5.html" class="sidebar-item-text sidebar-link">Assignment 5</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment6.html" class="sidebar-item-text sidebar-link">Assignment 6</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment7.html" class="sidebar-item-text sidebar-link">Assignment 7</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment8.html" class="sidebar-item-text sidebar-link">Assignment 8</a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./assignment9.html" class="sidebar-item-text sidebar-link">Assignment 9</a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#general-information" id="toc-general-information" class="nav-link active" data-scroll-target="#general-information"><span class="toc-section-number">1</span>  General information</a></li>
  <li><a href="#stan-warm-up-linear-model-of-bda-retention-with-stan" id="toc-stan-warm-up-linear-model-of-bda-retention-with-stan" class="nav-link" data-scroll-target="#stan-warm-up-linear-model-of-bda-retention-with-stan"><span class="toc-section-number">2</span>  Stan warm-up: linear model of BDA retention with Stan</a>
  <ul class="collapse">
  <li><a href="#data-preparation-and-sampling-from-the-posterior" id="toc-data-preparation-and-sampling-from-the-posterior" class="nav-link" data-scroll-target="#data-preparation-and-sampling-from-the-posterior"><span class="toc-section-number">2.1</span>  Data preparation and sampling from the posterior</a></li>
  <li><a href="#quick-check-for-sampling-convergence" id="toc-quick-check-for-sampling-convergence" class="nav-link" data-scroll-target="#quick-check-for-sampling-convergence"><span class="toc-section-number">2.2</span>  Quick check for sampling convergence</a></li>
  </ul></li>
  <li><a href="#diagnosing-problems-in-hmc-nuts-sampling" id="toc-diagnosing-problems-in-hmc-nuts-sampling" class="nav-link" data-scroll-target="#diagnosing-problems-in-hmc-nuts-sampling"><span class="toc-section-number">3</span>  Diagnosing Problems in HMC-NUTS Sampling</a>
  <ul class="collapse">
  <li><a href="#improper-posterior" id="toc-improper-posterior" class="nav-link" data-scroll-target="#improper-posterior"><span class="toc-section-number">3.1</span>  Improper Posterior</a>
  <ul class="collapse">
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data"><span class="toc-section-number">3.1.1</span>  Data</a></li>
  <li><a href="#compile-and-sample-from-the-following-model" id="toc-compile-and-sample-from-the-following-model" class="nav-link" data-scroll-target="#compile-and-sample-from-the-following-model"><span class="toc-section-number">3.1.2</span>  Compile and Sample from the following model</a></li>
  <li><a href="#convergence-diagnostics" id="toc-convergence-diagnostics" class="nav-link" data-scroll-target="#convergence-diagnostics"><span class="toc-section-number">3.1.3</span>  Convergence diagnostics</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#generalized-linear-model-bioassay-model-with-stan" id="toc-generalized-linear-model-bioassay-model-with-stan" class="nav-link" data-scroll-target="#generalized-linear-model-bioassay-model-with-stan"><span class="toc-section-number">4</span>  Generalized linear model: Bioassay model with Stan</a>
  <ul class="collapse">
  <li><a href="#run-model-and-analyse-output" id="toc-run-model-and-analyse-output" class="nav-link" data-scroll-target="#run-model-and-analyse-output"><span class="toc-section-number">4.1</span>  Run model and analyse output</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title d-none d-lg-block">Notebook for Assignment 6</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Aki Vehtari et al. </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="general-information" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> General information</h1>
<p>This assignment is related to Lecture 6 and Chapter 12.</p>
<p>We recommend using <a href="https://jupyter.cs.aalto.fi">JupyterHub</a> (which has all the needed packages pre-installed).</p>
<div class="hint">
<p><strong>Reading instructions:</strong></p>
<ul>
<li><a href="../BDA3_notes.html#ch12"><strong>The reading instructions for BDA3 Chapter 12</strong></a>.</li>
</ul>
</div>
<div class="callout-tip callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
General Instructions for Answering the Assignment Questions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<ul>
<li>Questions below are exact copies of the text found in the MyCourses quiz and should serve as a notebook where you can store notes and code.</li>
<li>We recommend opening these notebooks in the Aalto JupyterHub, see <a href="https://avehtari.github.io/BDA_course_Aalto/FAQ.html#How_to_use_R_and_RStudio_remotely"><strong>how to use R and RStudio remotely</strong></a>.</li>
<li>For inspiration for code, have a look at the <a href="https://avehtari.github.io/BDA_course_Aalto/demos.html"><strong>BDA R Demos</strong></a> and the specific Assignment code notebooks</li>
<li>Recommended additional self study exercises for each chapter in BDA3 are listed in the course web page. These will help to gain deeper understanding of the topic.</li>
<li>Common questions and answers regarding installation and technical problems can be found in <a href="https://avehtari.github.io/BDA_course_Aalto/FAQ.html">Frequently Asked Questions (FAQ)</a>.</li>
<li>Deadlines for all assignments can be found on the course web page and in MyCourses.</li>
<li>You are allowed to discuss assignments with your friends, but it is not allowed to copy solutions directly from other students or from internet.</li>
<li>Do not share your answers publicly.</li>
<li>Do not copy answers from the internet or from previous years. We compare the answers to the answers from previous years and to the answers from other students this year.</li>
<li>Use of AI is allowed on the course, but the most of the work needs to by the student, and you need to report whether you used AI and in which way you used them (See <a href="https://www.aalto.fi/en/services/guidance-for-the-use-of-artificial-intelligence-in-teaching-and-learning-at-aalto-university">points 5 and 6 in Aalto guidelines for use of AI in teaching</a>).</li>
<li>All suspected plagiarism will be reported and investigated. See more about the <a href="https://into.aalto.fi/display/ensaannot/Aalto+University+Code+of+Academic+Integrity+and+Handling+Violations+Thereof"><strong>Aalto University Code of Academic Integrity and Handling Violations Thereof</strong></a>.</li>
<li>If you have any suggestions or improvements to the course material, please post in the course chat feedback channel, create an issue, or submit a pull request to the public repository!</li>
</ul>
</div>
</div>
</div>
<div class="callout-warning callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Setup
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>The following installs and loads the <code>aaltobda</code> package:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(aaltobda)){</span>
<span id="cb1-2"><a href="#cb1-2"></a>    <span class="fu">install.packages</span>(<span class="st">"aaltobda"</span>, <span class="at">repos =</span> <span class="fu">c</span>(<span class="st">"https://avehtari.github.io/BDA_course_Aalto/"</span>, <span class="fu">getOption</span>(<span class="st">"repos"</span>)))</span>
<span id="cb1-3"><a href="#cb1-3"></a>    <span class="fu">library</span>(aaltobda)</span>
<span id="cb1-4"><a href="#cb1-4"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: aaltobda</code></pre>
</div>
</div>
<p>The following installs and loads the <a href="https://github.com/stefano-meschiari/latex2exp"><code>latex2exp</code> package</a>, which allows us to use LaTeX in plots:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(latex2exp)){</span>
<span id="cb3-2"><a href="#cb3-2"></a>    <span class="fu">install.packages</span>(<span class="st">"latex2exp"</span>)</span>
<span id="cb3-3"><a href="#cb3-3"></a>    <span class="fu">library</span>(latex2exp)</span>
<span id="cb3-4"><a href="#cb3-4"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: latex2exp</code></pre>
</div>
</div>
<p>The following installs and loads the <a href="https://github.com/stan-dev/posterior"><code>posterior</code> package</a> which imports the <code>rhat_basic()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(posterior)){</span>
<span id="cb5-2"><a href="#cb5-2"></a>    <span class="fu">install.packages</span>(<span class="st">"posterior"</span>)</span>
<span id="cb5-3"><a href="#cb5-3"></a>    <span class="fu">library</span>(posterior)</span>
<span id="cb5-4"><a href="#cb5-4"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: posterior</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>This is posterior version 1.4.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'posterior'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:aaltobda':

    mcse_quantile</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    mad, sd, var</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    %in%, match</code></pre>
</div>
</div>
<p>The following installs and loads the <a href="https://ggplot2.tidyverse.org/"><code>ggplot2</code> package</a> and the <a href="https://mc-stan.org/bayesplot/index.html"><code>bayesplot</code> package</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(ggplot2)){</span>
<span id="cb12-2"><a href="#cb12-2"></a>    <span class="fu">install.packages</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb12-3"><a href="#cb12-3"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb12-4"><a href="#cb12-4"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ggplot2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'ggplot2' was built under R version 4.3.1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(bayesplot)){</span>
<span id="cb15-2"><a href="#cb15-2"></a>    <span class="fu">install.packages</span>(<span class="st">"bayesplot"</span>)</span>
<span id="cb15-3"><a href="#cb15-3"></a>    <span class="fu">library</span>(bayesplot)</span>
<span id="cb15-4"><a href="#cb15-4"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: bayesplot</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>This is bayesplot version 1.10.0</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>- Online documentation and vignettes at mc-stan.org/bayesplot</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>- bayesplot theme set to bayesplot::theme_default()</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>   * Does _not_ affect other ggplot2 plots</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>   * See ?bayesplot_theme_set for details on theme setting</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'bayesplot'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:posterior':

    rhat</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(dplyr)){</span>
<span id="cb24-2"><a href="#cb24-2"></a>    <span class="fu">install.packages</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb24-3"><a href="#cb24-3"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb24-4"><a href="#cb24-4"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: dplyr</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(tidyr)){</span>
<span id="cb29-2"><a href="#cb29-2"></a>    <span class="fu">install.packages</span>(<span class="st">"tidyr"</span>)</span>
<span id="cb29-3"><a href="#cb29-3"></a>    <span class="fu">library</span>(tidyr)</span>
<span id="cb29-4"><a href="#cb29-4"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: tidyr</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="co"># Some additional set-up to make plots legible</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>ggplot2<span class="sc">::</span><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">14</span>))</span>
<span id="cb31-3"><a href="#cb31-3"></a>bayesplot<span class="sc">::</span><span class="fu">bayesplot_theme_set</span>(<span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">14</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The following installs and loads the <a href="https://mc-stan.org/cmdstanr/"><code>cmdstanr</code> package</a> and tries to install <code>cmdstan</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(cmdstanr)){</span>
<span id="cb32-2"><a href="#cb32-2"></a>    <span class="fu">install.packages</span>(<span class="st">"cmdstanr"</span>, <span class="at">repos =</span> <span class="fu">c</span>(<span class="st">"https://mc-stan.org/r-packages/"</span>, <span class="fu">getOption</span>(<span class="st">"repos"</span>)))</span>
<span id="cb32-3"><a href="#cb32-3"></a>    <span class="fu">library</span>(cmdstanr)</span>
<span id="cb32-4"><a href="#cb32-4"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: cmdstanr</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>This is cmdstanr version 0.5.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>- CmdStanR documentation and vignettes: mc-stan.org/cmdstanr</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>- CmdStan path: /Users/dk/.cmdstan/cmdstan-2.33.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>- CmdStan version: 2.33.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
A newer version of CmdStan is available. See ?install_cmdstan() to install it.
To disable this check set option or environment variable CMDSTANR_NO_VER_CHECK=TRUE.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a>cmdstan_installed <span class="ot">&lt;-</span> <span class="cf">function</span>(){</span>
<span id="cb39-2"><a href="#cb39-2"></a>  res <span class="ot">&lt;-</span> <span class="fu">try</span>(out <span class="ot">&lt;-</span> cmdstanr<span class="sc">::</span><span class="fu">cmdstan_path</span>(), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb39-3"><a href="#cb39-3"></a>  <span class="sc">!</span><span class="fu">inherits</span>(res, <span class="st">"try-error"</span>)</span>
<span id="cb39-4"><a href="#cb39-4"></a>}</span>
<span id="cb39-5"><a href="#cb39-5"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">cmdstan_installed</span>()){</span>
<span id="cb39-6"><a href="#cb39-6"></a>    <span class="fu">install_cmdstan</span>()</span>
<span id="cb39-7"><a href="#cb39-7"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="stan-warm-up-linear-model-of-bda-retention-with-stan" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Stan warm-up: linear model of BDA retention with Stan</h1>
<p>From 2018 to 2022, we have been keeping track of assignment submissions for the BDA course given the number of submissions for the 1st assignment. We will fit a simple linear model to answer two questions of interest:</p>
<ul>
<li>What is the trend of student retention as measured by assignment submissions?</li>
<li>Given the submission rates for assignments 1–8, how many students will complete the final 9th assignment (and potentially pass the course)?</li>
</ul>
<p>The author has given you the broken Stan code below, which they intend to encode the following linear model: <span class="math display">\[
\begin{aligned}
p(\alpha,\beta,\sigma) &amp;= \mathrm{const.}
      &amp; \text{(improper flat prior)}&amp;\text{ and}\\
p(y|x,\alpha,\beta,\sigma) &amp;= p_\mathrm{normal}(y|\alpha + \beta x, \sigma)
      &amp; \text{(normal likelihood)} &amp;\text{.}
\end{aligned}
\]</span> In both the statistical model above and in the Stan model below, <span class="math inline">\(x \in \mathbb{R}^N\)</span> and <span class="math inline">\(y \in \mathbb{R}^N\)</span> are vectors of the covariates / predictors (the assignment number) and vectors of the observation (proportions of students who have handed in the respective assignment). <span class="math inline">\(\alpha \in \mathbb{R}\)</span> is the unknown scalar intercept, <span class="math inline">\(\beta \in \mathbb{R}\)</span> is the unknown scalar slope and <span class="math inline">\(\sigma \in \mathbb{R}_{&gt;0}\)</span> is the unknown scalar observation standard deviation. The statistical model further implies <span class="math display">\[
p(y_\mathrm{pred.}|x_\mathrm{pred.},\alpha,\beta,\sigma) = p_\mathrm{normal}(y_\mathrm{pred.}|\alpha + \beta x_\mathrm{pred.}, \sigma)
\]</span> as the predictive distribution for a new observation <span class="math inline">\(y_\mathrm{pred.}\)</span> at a given new covariate value <span class="math inline">\(x_\mathrm{pred.}\)</span>.</p>
<p>You can download <a href="./additional_files/assignment6/linear_model.stan">the broken stan file from github</a>.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode numberSource stan number-lines code-with-copy"><code class="sourceCode stan"><span id="cb40-1"><a href="#cb40-1"></a><span class="kw">data</span> { <span class="co">#&lt;1&gt;</span></span>
<span id="cb40-2"><a href="#cb40-2"></a>    <span class="co">// number of data points</span></span>
<span id="cb40-3"><a href="#cb40-3"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; N; </span>
<span id="cb40-4"><a href="#cb40-4"></a>    <span class="co">// covariate / predictor</span></span>
<span id="cb40-5"><a href="#cb40-5"></a>    <span class="dt">vector</span>[N] x; </span>
<span id="cb40-6"><a href="#cb40-6"></a>    <span class="co">// observations</span></span>
<span id="cb40-7"><a href="#cb40-7"></a>    <span class="dt">vector</span>[N] y; </span>
<span id="cb40-8"><a href="#cb40-8"></a>    <span class="co">// number of covariate values to make predictions at</span></span>
<span id="cb40-9"><a href="#cb40-9"></a>    <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; no_predictions;</span>
<span id="cb40-10"><a href="#cb40-10"></a>    <span class="co">// covariate values to make predictions at</span></span>
<span id="cb40-11"><a href="#cb40-11"></a>    <span class="dt">vector</span>[no_predictions] x_predictions; </span>
<span id="cb40-12"><a href="#cb40-12"></a>} </span>
<span id="cb40-13"><a href="#cb40-13"></a><span class="kw">parameters</span> { <span class="co">#&lt;2&gt;</span></span>
<span id="cb40-14"><a href="#cb40-14"></a>    <span class="co">// intercept</span></span>
<span id="cb40-15"><a href="#cb40-15"></a>    <span class="dt">real</span> alpha; </span>
<span id="cb40-16"><a href="#cb40-16"></a>    <span class="co">// slope</span></span>
<span id="cb40-17"><a href="#cb40-17"></a>    <span class="dt">real</span> beta; </span>
<span id="cb40-18"><a href="#cb40-18"></a>    <span class="co">// the standard deviation should be constrained to be positive</span></span>
<span id="cb40-19"><a href="#cb40-19"></a>    <span class="dt">real</span>&lt;<span class="kw">upper</span>=<span class="dv">0</span>&gt; sigma; </span>
<span id="cb40-20"><a href="#cb40-20"></a>} </span>
<span id="cb40-21"><a href="#cb40-21"></a><span class="kw">transformed parameters</span> { <span class="co">#&lt;3&gt;</span></span>
<span id="cb40-22"><a href="#cb40-22"></a>    <span class="co">// deterministic transformation of parameters and data</span></span>
<span id="cb40-23"><a href="#cb40-23"></a>    <span class="dt">vector</span>[N] mu = alpha + beta * x <span class="co">// linear model</span></span>
<span id="cb40-24"><a href="#cb40-24"></a>} </span>
<span id="cb40-25"><a href="#cb40-25"></a><span class="kw">model</span> { <span class="co">#&lt;4&gt;</span></span>
<span id="cb40-26"><a href="#cb40-26"></a>    <span class="co">// observation model / likelihood</span></span>
<span id="cb40-27"><a href="#cb40-27"></a>    y ~ normal(mu, sigma); </span>
<span id="cb40-28"><a href="#cb40-28"></a>} </span>
<span id="cb40-29"><a href="#cb40-29"></a><span class="kw">generated quantities</span> { <span class="co">#&lt;5&gt;</span></span>
<span id="cb40-30"><a href="#cb40-30"></a>    <span class="co">// compute the means for the covariate values at which to make predictions</span></span>
<span id="cb40-31"><a href="#cb40-31"></a>    <span class="dt">vector</span>[no_predictions] mu_pred = alpha + beta * x_predictions;</span>
<span id="cb40-32"><a href="#cb40-32"></a>    <span class="co">// sample from the predictive distribution, a normal(mu_pred, sigma).</span></span>
<span id="cb40-33"><a href="#cb40-33"></a>    <span class="dt">array</span>[no_predictions] <span class="dt">real</span> y_pred = normal_rng(mu, sigma);</span>
<span id="cb40-34"><a href="#cb40-34"></a>} </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="hint">
<ol type="1">
<li><p>This is <a href="https://mc-stan.org/docs/reference-manual/program-block-data.html">Stan’s data block</a>: “<em>The <code>data</code> block is for the declaration of variables that are read in as data. […] Each variable’s value is validated against its declaration as it is read. For example, if a variable sigma is declared as <code>real&lt;lower=0&gt;</code>, then trying to assign it a negative value will raise an error. As a result, data type errors will be caught as early as possible. Similarly, attempts to provide data of the wrong size for a compound data structure will also raise an error.</em>” For more information, follow the link.</p></li>
<li><p>This is <a href="https://mc-stan.org/docs/reference-manual/program-block-parameters.html">Stan’s parameters block</a>: “<em>The variables declared in the <code>parameters</code> program block correspond directly to the variables being sampled by Stan’s samplers (HMC and NUTS). From a user’s perspective, the parameters in the program block are the parameters being sampled by Stan.</em>” For more information, follow the link.</p></li>
<li><p>This is <a href="https://mc-stan.org/docs/reference-manual/program-block-transformed-parameters.html">Stan’s transformed parameters block</a>: “<em>The <code>transformed parameters</code> program block consists of optional variable declarations followed by statements. After the statements are executed, the constraints on the transformed parameters are validated. Any variable declared as a transformed parameter is part of the output produced for draws.</em>” For more information, follow the link.</p></li>
<li><p>This is <a href="https://mc-stan.org/docs/reference-manual/program-block-model.html">Stan’s model block</a>: “<em>The <code>model</code> program block consists of optional variable declarations followed by statements. The variables in the <code>model</code> block are local variables and are not written as part of the output. […] The statements in the <code>model</code> block typically define the model. This is the block in which probability (sampling notation) statements are allowed.</em>” For more information, follow the link.</p></li>
<li><p>This is <a href="https://mc-stan.org/docs/reference-manual/program-block-generated-quantities.html">Stan’s generated quantities block</a>: “<em>The <code>generated quantities</code> program block is rather different than the other blocks. Nothing in the <code>generated quantities</code> block affects the sampled parameter values. The block is executed only after a sample has been generated.</em>” For more information, follow the link.</p></li>
</ol>
</div>
<section id="data-preparation-and-sampling-from-the-posterior" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="data-preparation-and-sampling-from-the-posterior"><span class="header-section-number">2.1</span> Data preparation and sampling from the posterior</h2>
<p><strong>Data assembly happens here</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="co"># These are our observations y: the proportion of students handing in each assignment (1-8),</span></span>
<span id="cb41-2"><a href="#cb41-2"></a><span class="co"># sorted by year (row-wise) and assignment (column-wise).</span></span>
<span id="cb41-3"><a href="#cb41-3"></a><span class="co"># While the code suggest a matrix structure, </span></span>
<span id="cb41-4"><a href="#cb41-4"></a><span class="co"># the result will actually be a vector of length N = no_years * no_assignments</span></span>
<span id="cb41-5"><a href="#cb41-5"></a>propstudents<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="fu">c</span>(<span class="dv">176</span>, <span class="dv">174</span>, <span class="dv">158</span>, <span class="dv">135</span>, <span class="dv">138</span>, <span class="dv">129</span>, <span class="dv">126</span>, <span class="dv">123</span>)<span class="sc">/</span><span class="dv">176</span>,</span>
<span id="cb41-6"><a href="#cb41-6"></a>                <span class="fu">c</span>(<span class="dv">242</span>, <span class="dv">212</span>, <span class="dv">184</span>, <span class="dv">177</span>, <span class="dv">174</span>, <span class="dv">172</span>, <span class="dv">163</span>, <span class="dv">156</span>)<span class="sc">/</span><span class="dv">242</span>,</span>
<span id="cb41-7"><a href="#cb41-7"></a>                <span class="fu">c</span>(<span class="dv">332</span>, <span class="dv">310</span>, <span class="dv">278</span>, <span class="dv">258</span>, <span class="dv">243</span>, <span class="dv">242</span>, <span class="dv">226</span>, <span class="dv">224</span>)<span class="sc">/</span><span class="dv">332</span>,</span>
<span id="cb41-8"><a href="#cb41-8"></a>                <span class="fu">c</span>(<span class="dv">301</span>, <span class="dv">269</span>, <span class="dv">231</span>, <span class="dv">232</span>, <span class="dv">217</span>, <span class="dv">208</span>, <span class="dv">193</span>, <span class="dv">191</span>)<span class="sc">/</span><span class="dv">301</span>,</span>
<span id="cb41-9"><a href="#cb41-9"></a>                <span class="fu">c</span>(<span class="dv">245</span>, <span class="dv">240</span>, <span class="dv">228</span>, <span class="dv">217</span>, <span class="dv">206</span>, <span class="dv">199</span>, <span class="dv">191</span>, <span class="dv">182</span>)<span class="sc">/</span><span class="dv">245</span>,</span>
<span id="cb41-10"><a href="#cb41-10"></a>                <span class="fu">c</span>(<span class="dv">264</span>, <span class="dv">249</span>, <span class="dv">215</span>, <span class="dv">221</span>, <span class="dv">215</span>, <span class="dv">206</span>, <span class="dv">192</span>, <span class="dv">186</span>)<span class="sc">/</span><span class="dv">264</span>)</span>
<span id="cb41-11"><a href="#cb41-11"></a><span class="co"># These are our predictors x: for each observation, the corresponding assignment number.</span></span>
<span id="cb41-12"><a href="#cb41-12"></a>assignment <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="dv">6</span>)</span>
<span id="cb41-13"><a href="#cb41-13"></a><span class="co"># These are in some sense our test data: the proportion of students handing in the last assignment (9),</span></span>
<span id="cb41-14"><a href="#cb41-14"></a><span class="co"># sorted by year. </span></span>
<span id="cb41-15"><a href="#cb41-15"></a><span class="co"># Usually, we would not want to split our data like that and instead</span></span>
<span id="cb41-16"><a href="#cb41-16"></a><span class="co"># use e.g. Leave-One-Out Cross-Validation (LOO-CV, see e.g. http://mc-stan.org/loo/index.html)</span></span>
<span id="cb41-17"><a href="#cb41-17"></a><span class="co"># to evaluate model performance.</span></span>
<span id="cb41-18"><a href="#cb41-18"></a>propstudents9 <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">121</span><span class="sc">/</span><span class="dv">176</span>, <span class="dv">153</span><span class="sc">/</span><span class="dv">242</span>, <span class="dv">218</span><span class="sc">/</span><span class="dv">332</span>, <span class="dv">190</span><span class="sc">/</span><span class="dv">301</span>, <span class="dv">175</span><span class="sc">/</span><span class="dv">245</span>, <span class="dv">179</span><span class="sc">/</span><span class="dv">264</span>)</span>
<span id="cb41-19"><a href="#cb41-19"></a><span class="co"># The total number of assignments</span></span>
<span id="cb41-20"><a href="#cb41-20"></a>no_assignments <span class="ot">=</span> <span class="dv">9</span></span>
<span id="cb41-21"><a href="#cb41-21"></a><span class="co"># The assignment numbers for which we want to generate predictions</span></span>
<span id="cb41-22"><a href="#cb41-22"></a>x_predictions <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span>no_assignments</span>
<span id="cb41-23"><a href="#cb41-23"></a><span class="co"># (Cmd)Stan(R) expects the data to be passed in the below format:</span></span>
<span id="cb41-24"><a href="#cb41-24"></a>model_data <span class="ot">=</span> <span class="fu">list</span>(<span class="at">N=</span><span class="fu">length</span>(assignment),</span>
<span id="cb41-25"><a href="#cb41-25"></a>                 <span class="at">x=</span>assignment,</span>
<span id="cb41-26"><a href="#cb41-26"></a>                 <span class="at">y=</span>propstudents,</span>
<span id="cb41-27"><a href="#cb41-27"></a>                 <span class="at">no_predictions=</span>no_assignments,</span>
<span id="cb41-28"><a href="#cb41-28"></a>                 <span class="at">x_predictions=</span>x_predictions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Sampling from the posterior distribution happens here</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a><span class="co"># This reads the file at the specified path and tries to compile it. </span></span>
<span id="cb42-2"><a href="#cb42-2"></a><span class="co"># If it fails, an error is thrown.</span></span>
<span id="cb42-3"><a href="#cb42-3"></a>retention_model <span class="ot">=</span> <span class="fu">cmdstan_model</span>(<span class="st">"assignment6_linear_model.stan"</span>)</span>
<span id="cb42-4"><a href="#cb42-4"></a><span class="co"># This "out &lt;- capture.output(...)" construction suppresses output from cmdstanr</span></span>
<span id="cb42-5"><a href="#cb42-5"></a><span class="co"># See also https://github.com/stan-dev/cmdstanr/issues/646</span></span>
<span id="cb42-6"><a href="#cb42-6"></a>out <span class="ot">&lt;-</span> <span class="fu">capture.output</span>(</span>
<span id="cb42-7"><a href="#cb42-7"></a>    <span class="co"># Sampling from the posterior distribution happens here:</span></span>
<span id="cb42-8"><a href="#cb42-8"></a>    fit <span class="ot">&lt;-</span> retention_model<span class="sc">$</span><span class="fu">sample</span>(<span class="at">data=</span>model_data, <span class="at">refresh=</span><span class="dv">0</span>, <span class="at">show_messages=</span><span class="cn">FALSE</span>)</span>
<span id="cb42-9"><a href="#cb42-9"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Draws postprocessing happens here</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="co"># This extracts the draws from the sampling result as a data.frame.</span></span>
<span id="cb43-2"><a href="#cb43-2"></a>draws_df <span class="ot">=</span> fit<span class="sc">$</span><span class="fu">draws</span>(<span class="at">format=</span><span class="st">"draws_df"</span>)</span>
<span id="cb43-3"><a href="#cb43-3"></a></span>
<span id="cb43-4"><a href="#cb43-4"></a><span class="co"># This does some data/draws wrangling to compute the 5, 50 and 95 percentiles of </span></span>
<span id="cb43-5"><a href="#cb43-5"></a><span class="co"># the mean at the specified covariate values (x_predictions). </span></span>
<span id="cb43-6"><a href="#cb43-6"></a><span class="co"># It can be instructive to play around with each of the data processing steps</span></span>
<span id="cb43-7"><a href="#cb43-7"></a><span class="co"># to find out what each step does, e.g. by removing parts from the back like "|&gt;  gather(pct,y,-x)"</span></span>
<span id="cb43-8"><a href="#cb43-8"></a><span class="co"># and printing the resulting data.frame.</span></span>
<span id="cb43-9"><a href="#cb43-9"></a>mu_quantiles_df <span class="ot">=</span> draws_df <span class="sc">|&gt;</span> </span>
<span id="cb43-10"><a href="#cb43-10"></a>      <span class="fu">subset_draws</span>(<span class="at">variable =</span> <span class="fu">c</span>(<span class="st">"mu_pred"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb43-11"><a href="#cb43-11"></a>      <span class="fu">summarise_draws</span>(<span class="sc">~</span><span class="fu">quantile2</span>(.x, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.05</span>, .<span class="dv">5</span>, <span class="fl">0.95</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb43-12"><a href="#cb43-12"></a>      <span class="fu">mutate</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>) <span class="sc">|&gt;</span> </span>
<span id="cb43-13"><a href="#cb43-13"></a>      <span class="fu">pivot_longer</span>(<span class="fu">c</span>(q5, q50, q95), <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"pct"</span>))</span>
<span id="cb43-14"><a href="#cb43-14"></a><span class="co"># Same as above, but for the predictions.</span></span>
<span id="cb43-15"><a href="#cb43-15"></a>y_quantiles_df <span class="ot">=</span> draws_df <span class="sc">|&gt;</span> </span>
<span id="cb43-16"><a href="#cb43-16"></a>      <span class="fu">subset_draws</span>(<span class="at">variable =</span> <span class="fu">c</span>(<span class="st">"y_pred"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb43-17"><a href="#cb43-17"></a>      <span class="fu">summarise_draws</span>(<span class="sc">~</span><span class="fu">quantile2</span>(.x, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.05</span>, .<span class="dv">5</span>, <span class="fl">0.95</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb43-18"><a href="#cb43-18"></a>      <span class="fu">mutate</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>) <span class="sc">|&gt;</span> </span>
<span id="cb43-19"><a href="#cb43-19"></a>      <span class="fu">pivot_longer</span>(<span class="fu">c</span>(q5, q50, q95), <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"pct"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Plotting happens here</strong>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb44-2"><a href="#cb44-2"></a>  <span class="co"># scatter plot of the training data:  </span></span>
<span id="cb44-3"><a href="#cb44-3"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb44-4"><a href="#cb44-4"></a>    <span class="fu">aes</span>(x, y, <span class="at">color=</span>assignment), </span>
<span id="cb44-5"><a href="#cb44-5"></a>    <span class="at">data=</span><span class="fu">data.frame</span>(<span class="at">x=</span>assignment, <span class="at">y=</span>propstudents, <span class="at">assignment=</span><span class="st">"1-8"</span>)</span>
<span id="cb44-6"><a href="#cb44-6"></a>) <span class="sc">+</span></span>
<span id="cb44-7"><a href="#cb44-7"></a>  <span class="co"># scatter plot of the test data:</span></span>
<span id="cb44-8"><a href="#cb44-8"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb44-9"><a href="#cb44-9"></a>    <span class="fu">aes</span>(x, y, <span class="at">color=</span>assignment), </span>
<span id="cb44-10"><a href="#cb44-10"></a>    <span class="at">data=</span><span class="fu">data.frame</span>(<span class="at">x=</span>no_assignments, <span class="at">y=</span>propstudents9, <span class="at">assignment=</span><span class="st">"9"</span>)</span>
<span id="cb44-11"><a href="#cb44-11"></a>) <span class="sc">+</span></span>
<span id="cb44-12"><a href="#cb44-12"></a>  <span class="co"># you have to tell us what this plots:</span></span>
<span id="cb44-13"><a href="#cb44-13"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(x,<span class="at">y=</span>value,<span class="at">linetype=</span>pct), <span class="at">data=</span>mu_quantiles_df, <span class="at">color=</span><span class="st">'grey'</span>, <span class="at">linewidth=</span><span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb44-14"><a href="#cb44-14"></a>  <span class="co"># you have to tell us what this plots:</span></span>
<span id="cb44-15"><a href="#cb44-15"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(x,<span class="at">y=</span>value,<span class="at">linetype=</span>pct), <span class="at">data=</span>y_quantiles_df, <span class="at">color=</span><span class="st">'red'</span>) <span class="sc">+</span></span>
<span id="cb44-16"><a href="#cb44-16"></a>  <span class="co"># adding xticks for each assignment:</span></span>
<span id="cb44-17"><a href="#cb44-17"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="dv">1</span><span class="sc">:</span>no_assignments) <span class="sc">+</span></span>
<span id="cb44-18"><a href="#cb44-18"></a>  <span class="co"># adding labels to the plot:</span></span>
<span id="cb44-19"><a href="#cb44-19"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"assignment submission %"</span>, <span class="at">x=</span><span class="st">"assignment number"</span>) <span class="sc">+</span></span>
<span id="cb44-20"><a href="#cb44-20"></a>  <span class="co"># specifying that line types repeat:</span></span>
<span id="cb44-21"><a href="#cb44-21"></a>  <span class="fu">scale_linetype_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb44-22"><a href="#cb44-22"></a>  <span class="co"># Specify colours of the observations:</span></span>
<span id="cb44-23"><a href="#cb44-23"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"1-8"</span><span class="ot">=</span><span class="st">"black"</span>, <span class="st">"9"</span><span class="ot">=</span><span class="st">"blue"</span>)) <span class="sc">+</span></span>
<span id="cb44-24"><a href="#cb44-24"></a>  <span class="co"># remove the legend for the linetypes:</span></span>
<span id="cb44-25"><a href="#cb44-25"></a>  <span class="fu">guides</span>(<span class="at">linetype=</span><span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quick-check-for-sampling-convergence" class="level2 hint" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="quick-check-for-sampling-convergence"><span class="header-section-number">2.2</span> Quick check for sampling convergence</h2>
<p>If your model is correctly implemented, sampling from the posterior distribution should have been successful. You can check whether Stan thinks that sampling succeeded by inspecting the output of the below command, which you should be able to interpret with a little help from the <a href="https://mc-stan.org/docs/cmdstan-guide/diagnose.html">CmdStan User’s Guide</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a>fit<span class="sc">$</span><span class="fu">cmdstan_diagnose</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="diagnosing-problems-in-hmc-nuts-sampling" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Diagnosing Problems in HMC-NUTS Sampling</h1>
<section id="improper-posterior" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="improper-posterior"><span class="header-section-number">3.1</span> Improper Posterior</h2>
<p>An unbounded likelihood without a proper prior can lead to an improper posterior. We recommend to always use proper priors (integral over a proper distribution is finite) to guarantee proper posteriors.</p>
<p>A commonly used model that can have unbounded likelihood is logistic regression with complete separation in data.</p>
<section id="data" class="level3" data-number="3.1.1">
<h3 data-number="3.1.1" class="anchored" data-anchor-id="data"><span class="header-section-number">3.1.1</span> Data</h3>
<p>Univariate continous predictor <span class="math inline">\(x\)</span>, binary target <span class="math inline">\(y\)</span>, and the two classes are completely separable, which leads to unbounded likelihood.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1"></a><span class="fu">set.seed</span>(<span class="dv">48927</span><span class="sc">+</span><span class="dv">4</span>)</span>
<span id="cb46-2"><a href="#cb46-2"></a>M<span class="ot">=</span><span class="dv">1</span>;</span>
<span id="cb46-3"><a href="#cb46-3"></a>N<span class="ot">=</span><span class="dv">10</span>;</span>
<span id="cb46-4"><a href="#cb46-4"></a>x<span class="ot">=</span><span class="fu">matrix</span>(<span class="fu">sort</span>(<span class="fu">rnorm</span>(N)),<span class="at">ncol=</span>M)</span>
<span id="cb46-5"><a href="#cb46-5"></a>y<span class="ot">=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">each=</span>N<span class="sc">/</span><span class="dv">2</span>)</span>
<span id="cb46-6"><a href="#cb46-6"></a>data_logit <span class="ot">&lt;-</span><span class="fu">list</span>(<span class="at">M =</span> M, <span class="at">N =</span> N, <span class="at">x =</span> x, <span class="at">y =</span> y)</span>
<span id="cb46-7"><a href="#cb46-7"></a><span class="fu">data.frame</span>(data_logit) <span class="sc">|&gt;</span></span>
<span id="cb46-8"><a href="#cb46-8"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(x, y)) <span class="sc">+</span></span>
<span id="cb46-9"><a href="#cb46-9"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>, <span class="at">shape=</span><span class="dv">1</span>, <span class="at">alpha=</span><span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb46-10"><a href="#cb46-10"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="template6_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="compile-and-sample-from-the-following-model" class="level3" data-number="3.1.2">
<h3 data-number="3.1.2" class="anchored" data-anchor-id="compile-and-sample-from-the-following-model"><span class="header-section-number">3.1.2</span> Compile and Sample from the following model</h3>
<div class="sourceCode" id="cb47"><pre class="sourceCode numberSource stan number-lines code-with-copy"><code class="sourceCode stan"><span id="cb47-1"><a href="#cb47-1"></a><span class="co">// logistic regression</span></span>
<span id="cb47-2"><a href="#cb47-2"></a><span class="kw">data</span> {</span>
<span id="cb47-3"><a href="#cb47-3"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; N;</span>
<span id="cb47-4"><a href="#cb47-4"></a>  <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>&gt; M;</span>
<span id="cb47-5"><a href="#cb47-5"></a>  <span class="dt">array</span>[N] <span class="dt">int</span>&lt;<span class="kw">lower</span>=<span class="dv">0</span>,<span class="kw">upper</span>=<span class="dv">1</span>&gt; y;</span>
<span id="cb47-6"><a href="#cb47-6"></a>  <span class="dt">matrix</span>[N,M] x;</span>
<span id="cb47-7"><a href="#cb47-7"></a>}</span>
<span id="cb47-8"><a href="#cb47-8"></a><span class="kw">parameters</span> {</span>
<span id="cb47-9"><a href="#cb47-9"></a>  <span class="dt">real</span> alpha;</span>
<span id="cb47-10"><a href="#cb47-10"></a>  <span class="dt">vector</span>[M] beta;</span>
<span id="cb47-11"><a href="#cb47-11"></a>}</span>
<span id="cb47-12"><a href="#cb47-12"></a><span class="kw">model</span> {</span>
<span id="cb47-13"><a href="#cb47-13"></a>  y ~ bernoulli_logit_glm(x, alpha, beta);</span>
<span id="cb47-14"><a href="#cb47-14"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="convergence-diagnostics" class="level3" data-number="3.1.3">
<h3 data-number="3.1.3" class="anchored" data-anchor-id="convergence-diagnostics"><span class="header-section-number">3.1.3</span> Convergence diagnostics</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1"></a><span class="co"># Get the summary diagnostics using [fit object name]]$diagnostic_summary()</span></span>
<span id="cb48-2"><a href="#cb48-2"></a></span>
<span id="cb48-3"><a href="#cb48-3"></a><span class="co"># Get summary statistics from your chains using summarize_draws() (hint: look at how we used the function last week)</span></span>
<span id="cb48-4"><a href="#cb48-4"></a></span>
<span id="cb48-5"><a href="#cb48-5"></a><span class="co"># Plot histograms and bivariate scatter plots using mcmc_pairs(). See for guidance on how to use the function here: https://mc-stan.org/bayesplot/reference/MCMC-scatterplots.html</span></span>
<span id="cb48-6"><a href="#cb48-6"></a></span>
<span id="cb48-7"><a href="#cb48-7"></a><span class="co"># To plot the tracplot of individual chains, use the mcmc_trace() function. See for guidance on how to use the function here: https://mc-stan.org/bayesplot/reference/MCMC-traces.html </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="generalized-linear-model-bioassay-model-with-stan" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Generalized linear model: Bioassay model with Stan</h1>
<p>Replicate the computations for the bioassay example of section 3.7 (BDA3) using Stan.</p>
<p>Write down the model for the bioassay data in Stan syntax. Use the slightly modified Gaussian prior from Assignment 4 and 5, that is <span class="math display">\[
    \alpha \sim normal(0,2), \quad
    \beta \sim normal(10,10)
\]</span></p>
<p>Write your Stan code here</p>
<section id="run-model-and-analyse-output" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="run-model-and-analyse-output"><span class="header-section-number">4.1</span> Run model and analyse output</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="co"># Load data</span></span>
<span id="cb49-2"><a href="#cb49-2"></a>data <span class="ot">=</span><span class="fu">list</span>(<span class="at">n =</span> bioassay<span class="sc">$</span>n, <span class="at">x =</span> bioassay<span class="sc">$</span>x, <span class="at">y =</span> bioassay<span class="sc">$</span>y, <span class="at">N =</span> <span class="fu">nrow</span>(bioassay))</span>
<span id="cb49-3"><a href="#cb49-3"></a></span>
<span id="cb49-4"><a href="#cb49-4"></a><span class="co"># Compile Model</span></span>
<span id="cb49-5"><a href="#cb49-5"></a>mod <span class="ot">&lt;-</span> <span class="fu">cmdstan_model</span>(<span class="co"># your stan model location)</span></span>
<span id="cb49-6"><a href="#cb49-6"></a></span>
<span id="cb49-7"><a href="#cb49-7"></a><span class="co"># Estimate model</span></span>
<span id="cb49-8"><a href="#cb49-8"></a>out <span class="ot">&lt;-</span> <span class="fu">capture.output</span>(</span>
<span id="cb49-9"><a href="#cb49-9"></a>  <span class="co"># Sampling from the posterior distribution happens here:</span></span>
<span id="cb49-10"><a href="#cb49-10"></a>  fit <span class="ot">&lt;-</span> mod<span class="sc">$</span><span class="fu">sample</span>(<span class="at">data=</span>data, <span class="at">refresh=</span><span class="dv">0</span>, <span class="at">show_messages=</span><span class="cn">FALSE</span>, <span class="at">seed =</span> <span class="dv">4711</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">iter_warmup =</span> <span class="dv">1000</span>, </span>
<span id="cb49-11"><a href="#cb49-11"></a>                    <span class="at">iter_sampling =</span> <span class="dv">3000</span>)</span>
<span id="cb49-12"><a href="#cb49-12"></a>)</span>
<span id="cb49-13"><a href="#cb49-13"></a></span>
<span id="cb49-14"><a href="#cb49-14"></a><span class="co"># This extracts the draws from the sampling result as a data.frame.</span></span>
<span id="cb49-15"><a href="#cb49-15"></a><span class="at">draws_df =</span> fit<span class="sc">$</span><span class="fu">draws</span>(<span class="at">format=</span><span class="st">"draws_df"</span>)</span>
<span id="cb49-16"><a href="#cb49-16"></a></span>
<span id="cb49-17"><a href="#cb49-17"></a><span class="co"># This is what you'll need for convergence diagnostics</span></span>
<span id="cb49-18"><a href="#cb49-18"></a><span class="fu">summarise_draws</span>(draws_df, <span class="at">Rhat=</span>rhat_basic, <span class="at">ESS=</span> ess_mean, <span class="sc">~</span><span class="fu">ess_quantile</span>(.x, <span class="at">probs =</span> <span class="fl">0.05</span>))</span>
<span id="cb49-19"><a href="#cb49-19"></a></span>
<span id="cb49-20"><a href="#cb49-20"></a><span class="co"># Scatter plot of the draws</span></span>
<span id="cb49-21"><a href="#cb49-21"></a><span class="fu">mcmc_scatter</span>(draws_df, <span class="at">pars=</span><span class="fu">c</span>(<span class="st">"alpha"</span>, <span class="st">"beta"</span>))</span>
<span id="cb49-22"><a href="#cb49-22"></a></span>
<span id="cb49-23"><a href="#cb49-23"></a><span class="co"># Plot the autocorrelation function and compare to last week</span></span>
<span id="cb49-24"><a href="#cb49-24"></a><span class="fu">mcmc_acf</span>(draws_df, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"alpha"</span>,<span class="st">"beta"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb50" data-shortcodes="false"><pre class="sourceCode numberSource markdown number-lines code-with-copy"><code class="sourceCode markdown"><span id="cb50-1"><a href="#cb50-1"></a><span class="co">---</span></span>
<span id="cb50-2"><a href="#cb50-2"></a><span class="an">title:</span><span class="co"> "Notebook for Assignment 6"</span></span>
<span id="cb50-3"><a href="#cb50-3"></a><span class="an">author:</span><span class="co"> "Aki Vehtari et al."</span></span>
<span id="cb50-4"><a href="#cb50-4"></a><span class="an">format:</span></span>
<span id="cb50-5"><a href="#cb50-5"></a><span class="co">  html:</span></span>
<span id="cb50-6"><a href="#cb50-6"></a><span class="co">    toc: true</span></span>
<span id="cb50-7"><a href="#cb50-7"></a><span class="co">    code-tools: true</span></span>
<span id="cb50-8"><a href="#cb50-8"></a><span class="co">    code-line-numbers: true</span></span>
<span id="cb50-9"><a href="#cb50-9"></a><span class="co">    number-sections: true</span></span>
<span id="cb50-10"><a href="#cb50-10"></a><span class="co">    mainfont: Georgia, serif</span></span>
<span id="cb50-11"><a href="#cb50-11"></a><span class="an">editor:</span><span class="co"> source </span></span>
<span id="cb50-12"><a href="#cb50-12"></a><span class="co">---</span></span>
<span id="cb50-13"><a href="#cb50-13"></a></span>
<span id="cb50-14"><a href="#cb50-14"></a><span class="fu"># General information</span></span>
<span id="cb50-15"><a href="#cb50-15"></a></span>
<span id="cb50-16"><a href="#cb50-16"></a>This assignment is related to Lecture 6 and Chapter 12.</span>
<span id="cb50-17"><a href="#cb50-17"></a></span>
<span id="cb50-18"><a href="#cb50-18"></a>We recommend using <span class="co">[</span><span class="ot">JupyterHub</span><span class="co">](https://jupyter.cs.aalto.fi)</span> (which has all the needed packages pre-installed).</span>
<span id="cb50-19"><a href="#cb50-19"></a></span>
<span id="cb50-20"><a href="#cb50-20"></a>::: hint</span>
<span id="cb50-21"><a href="#cb50-21"></a>**Reading instructions:**</span>
<span id="cb50-22"><a href="#cb50-22"></a></span>
<span id="cb50-23"><a href="#cb50-23"></a><span class="ss">-   </span><span class="co">[</span><span class="ot">**The reading instructions for BDA3 Chapter 12**</span><span class="co">](../BDA3_notes.html#ch12)</span>.</span>
<span id="cb50-24"><a href="#cb50-24"></a>:::</span>
<span id="cb50-25"><a href="#cb50-25"></a></span>
<span id="cb50-26"><a href="#cb50-26"></a>{{&lt; include includes/_general_cloze_instructions.md &gt;}}</span>
<span id="cb50-27"><a href="#cb50-27"></a></span>
<span id="cb50-28"><a href="#cb50-28"></a>::: {.callout-warning collapse="false"}</span>
<span id="cb50-29"><a href="#cb50-29"></a><span class="fu">## Setup</span></span>
<span id="cb50-30"><a href="#cb50-30"></a></span>
<span id="cb50-31"><a href="#cb50-31"></a>The following installs and loads the <span class="in">`aaltobda`</span> package:</span>
<span id="cb50-32"><a href="#cb50-32"></a></span>
<span id="cb50-35"><a href="#cb50-35"></a><span class="in">```{r}</span></span>
<span id="cb50-36"><a href="#cb50-36"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(aaltobda)){</span>
<span id="cb50-37"><a href="#cb50-37"></a>    <span class="fu">install.packages</span>(<span class="st">"aaltobda"</span>, <span class="at">repos =</span> <span class="fu">c</span>(<span class="st">"https://avehtari.github.io/BDA_course_Aalto/"</span>, <span class="fu">getOption</span>(<span class="st">"repos"</span>)))</span>
<span id="cb50-38"><a href="#cb50-38"></a>    <span class="fu">library</span>(aaltobda)</span>
<span id="cb50-39"><a href="#cb50-39"></a>}</span>
<span id="cb50-40"><a href="#cb50-40"></a><span class="in">```</span></span>
<span id="cb50-41"><a href="#cb50-41"></a></span>
<span id="cb50-42"><a href="#cb50-42"></a>The following installs and loads the <span class="co">[</span><span class="ot">`latex2exp` package</span><span class="co">](https://github.com/stefano-meschiari/latex2exp)</span>, which allows us to use LaTeX in plots:</span>
<span id="cb50-43"><a href="#cb50-43"></a></span>
<span id="cb50-46"><a href="#cb50-46"></a><span class="in">```{r}</span></span>
<span id="cb50-47"><a href="#cb50-47"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(latex2exp)){</span>
<span id="cb50-48"><a href="#cb50-48"></a>    <span class="fu">install.packages</span>(<span class="st">"latex2exp"</span>)</span>
<span id="cb50-49"><a href="#cb50-49"></a>    <span class="fu">library</span>(latex2exp)</span>
<span id="cb50-50"><a href="#cb50-50"></a>}</span>
<span id="cb50-51"><a href="#cb50-51"></a><span class="in">```</span></span>
<span id="cb50-52"><a href="#cb50-52"></a></span>
<span id="cb50-53"><a href="#cb50-53"></a>The following installs and loads the <span class="co">[</span><span class="ot">`posterior` package</span><span class="co">](https://github.com/stan-dev/posterior)</span> which imports the <span class="in">`rhat_basic()`</span> function:</span>
<span id="cb50-54"><a href="#cb50-54"></a></span>
<span id="cb50-57"><a href="#cb50-57"></a><span class="in">```{r}</span></span>
<span id="cb50-58"><a href="#cb50-58"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(posterior)){</span>
<span id="cb50-59"><a href="#cb50-59"></a>    <span class="fu">install.packages</span>(<span class="st">"posterior"</span>)</span>
<span id="cb50-60"><a href="#cb50-60"></a>    <span class="fu">library</span>(posterior)</span>
<span id="cb50-61"><a href="#cb50-61"></a>}</span>
<span id="cb50-62"><a href="#cb50-62"></a><span class="in">```</span></span>
<span id="cb50-63"><a href="#cb50-63"></a></span>
<span id="cb50-64"><a href="#cb50-64"></a>The following installs and loads the <span class="co">[</span><span class="ot">`ggplot2` package</span><span class="co">](https://ggplot2.tidyverse.org/)</span> and the <span class="co">[</span><span class="ot">`bayesplot` package</span><span class="co">](https://mc-stan.org/bayesplot/index.html)</span></span>
<span id="cb50-65"><a href="#cb50-65"></a></span>
<span id="cb50-68"><a href="#cb50-68"></a><span class="in">```{r}</span></span>
<span id="cb50-69"><a href="#cb50-69"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(ggplot2)){</span>
<span id="cb50-70"><a href="#cb50-70"></a>    <span class="fu">install.packages</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb50-71"><a href="#cb50-71"></a>    <span class="fu">library</span>(ggplot2)</span>
<span id="cb50-72"><a href="#cb50-72"></a>}</span>
<span id="cb50-73"><a href="#cb50-73"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(bayesplot)){</span>
<span id="cb50-74"><a href="#cb50-74"></a>    <span class="fu">install.packages</span>(<span class="st">"bayesplot"</span>)</span>
<span id="cb50-75"><a href="#cb50-75"></a>    <span class="fu">library</span>(bayesplot)</span>
<span id="cb50-76"><a href="#cb50-76"></a>}</span>
<span id="cb50-77"><a href="#cb50-77"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(dplyr)){</span>
<span id="cb50-78"><a href="#cb50-78"></a>    <span class="fu">install.packages</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb50-79"><a href="#cb50-79"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb50-80"><a href="#cb50-80"></a>}</span>
<span id="cb50-81"><a href="#cb50-81"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(tidyr)){</span>
<span id="cb50-82"><a href="#cb50-82"></a>    <span class="fu">install.packages</span>(<span class="st">"tidyr"</span>)</span>
<span id="cb50-83"><a href="#cb50-83"></a>    <span class="fu">library</span>(tidyr)</span>
<span id="cb50-84"><a href="#cb50-84"></a>}</span>
<span id="cb50-85"><a href="#cb50-85"></a><span class="co"># Some additional set-up to make plots legible</span></span>
<span id="cb50-86"><a href="#cb50-86"></a>ggplot2<span class="sc">::</span><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">14</span>))</span>
<span id="cb50-87"><a href="#cb50-87"></a>bayesplot<span class="sc">::</span><span class="fu">bayesplot_theme_set</span>(<span class="fu">theme_minimal</span>(<span class="at">base_size =</span> <span class="dv">14</span>))</span>
<span id="cb50-88"><a href="#cb50-88"></a><span class="in">```</span></span>
<span id="cb50-89"><a href="#cb50-89"></a></span>
<span id="cb50-90"><a href="#cb50-90"></a>The following installs and loads the <span class="co">[</span><span class="ot">`cmdstanr` package</span><span class="co">](https://mc-stan.org/cmdstanr/)</span> and tries to install <span class="in">`cmdstan`</span>.</span>
<span id="cb50-91"><a href="#cb50-91"></a></span>
<span id="cb50-94"><a href="#cb50-94"></a><span class="in">```{r}</span></span>
<span id="cb50-95"><a href="#cb50-95"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">require</span>(cmdstanr)){</span>
<span id="cb50-96"><a href="#cb50-96"></a>    <span class="fu">install.packages</span>(<span class="st">"cmdstanr"</span>, <span class="at">repos =</span> <span class="fu">c</span>(<span class="st">"https://mc-stan.org/r-packages/"</span>, <span class="fu">getOption</span>(<span class="st">"repos"</span>)))</span>
<span id="cb50-97"><a href="#cb50-97"></a>    <span class="fu">library</span>(cmdstanr)</span>
<span id="cb50-98"><a href="#cb50-98"></a>}</span>
<span id="cb50-99"><a href="#cb50-99"></a>cmdstan_installed <span class="ot">&lt;-</span> <span class="cf">function</span>(){</span>
<span id="cb50-100"><a href="#cb50-100"></a>  res <span class="ot">&lt;-</span> <span class="fu">try</span>(out <span class="ot">&lt;-</span> cmdstanr<span class="sc">::</span><span class="fu">cmdstan_path</span>(), <span class="at">silent =</span> <span class="cn">TRUE</span>)</span>
<span id="cb50-101"><a href="#cb50-101"></a>  <span class="sc">!</span><span class="fu">inherits</span>(res, <span class="st">"try-error"</span>)</span>
<span id="cb50-102"><a href="#cb50-102"></a>}</span>
<span id="cb50-103"><a href="#cb50-103"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">cmdstan_installed</span>()){</span>
<span id="cb50-104"><a href="#cb50-104"></a>    <span class="fu">install_cmdstan</span>()</span>
<span id="cb50-105"><a href="#cb50-105"></a>}</span>
<span id="cb50-106"><a href="#cb50-106"></a><span class="in">```</span></span>
<span id="cb50-107"><a href="#cb50-107"></a>:::</span>
<span id="cb50-108"><a href="#cb50-108"></a></span>
<span id="cb50-109"><a href="#cb50-109"></a><span class="fu"># Stan warm-up: linear model of BDA retention with Stan</span></span>
<span id="cb50-110"><a href="#cb50-110"></a></span>
<span id="cb50-111"><a href="#cb50-111"></a>From 2018 to 2022, we have been keeping track of assignment submissions for the BDA course given the number of submissions for the 1st assignment. We will fit a simple linear model to answer two questions of interest:</span>
<span id="cb50-112"><a href="#cb50-112"></a></span>
<span id="cb50-113"><a href="#cb50-113"></a><span class="ss">-   </span>What is the trend of student retention as measured by assignment submissions?</span>
<span id="cb50-114"><a href="#cb50-114"></a><span class="ss">-   </span>Given the submission rates for assignments 1--8, how many students will complete the final 9th assignment (and potentially pass the course)?</span>
<span id="cb50-115"><a href="#cb50-115"></a></span>
<span id="cb50-116"><a href="#cb50-116"></a>The author has given you the broken Stan code below, which they intend to encode the following linear model: $$</span>
<span id="cb50-117"><a href="#cb50-117"></a>\begin{aligned}</span>
<span id="cb50-118"><a href="#cb50-118"></a>p(\alpha,\beta,\sigma) &amp;= \mathrm{const.}</span>
<span id="cb50-119"><a href="#cb50-119"></a>      &amp; \text{(improper flat prior)}&amp;\text{ and}<span class="sc">\\</span></span>
<span id="cb50-120"><a href="#cb50-120"></a>p(y|x,\alpha,\beta,\sigma) &amp;= p_\mathrm{normal}(y|\alpha + \beta x, \sigma)</span>
<span id="cb50-121"><a href="#cb50-121"></a>      &amp; \text{(normal likelihood)} &amp;\text{.}</span>
<span id="cb50-122"><a href="#cb50-122"></a>\end{aligned}</span>
<span id="cb50-123"><a href="#cb50-123"></a>$$ In both the statistical model above and in the Stan model below, $x \in \mathbb{R}^N$ and $y \in \mathbb{R}^N$ are vectors of the covariates / predictors (the assignment number) and vectors of the observation (proportions of students who have handed in the respective assignment). $\alpha \in \mathbb{R}$ is the unknown scalar intercept, $\beta \in \mathbb{R}$ is the unknown scalar slope and $\sigma \in \mathbb{R}_{&gt;0}$ is the unknown scalar observation standard deviation. The statistical model further implies $$</span>
<span id="cb50-124"><a href="#cb50-124"></a>p(y_\mathrm{pred.}|x_\mathrm{pred.},\alpha,\beta,\sigma) = p_\mathrm{normal}(y_\mathrm{pred.}|\alpha + \beta x_\mathrm{pred.}, \sigma)</span>
<span id="cb50-125"><a href="#cb50-125"></a>$$ as the predictive distribution for a new observation $y_\mathrm{pred.}$ at a given new covariate value $x_\mathrm{pred.}$.</span>
<span id="cb50-126"><a href="#cb50-126"></a></span>
<span id="cb50-127"><a href="#cb50-127"></a>You can download <span class="co">[</span><span class="ot">the broken stan file from github</span><span class="co">](./additional_files/assignment6/linear_model.stan)</span>.</span>
<span id="cb50-128"><a href="#cb50-128"></a></span>
<span id="cb50-129"><a href="#cb50-129"></a><span class="in">``` stan</span></span>
<span id="cb50-130"><a href="#cb50-130"></a><span class="in">data { #&lt;1&gt;</span></span>
<span id="cb50-131"><a href="#cb50-131"></a><span class="in">    // number of data points</span></span>
<span id="cb50-132"><a href="#cb50-132"></a><span class="in">    int&lt;lower=0&gt; N; </span></span>
<span id="cb50-133"><a href="#cb50-133"></a><span class="in">    // covariate / predictor</span></span>
<span id="cb50-134"><a href="#cb50-134"></a><span class="in">    vector[N] x; </span></span>
<span id="cb50-135"><a href="#cb50-135"></a><span class="in">    // observations</span></span>
<span id="cb50-136"><a href="#cb50-136"></a><span class="in">    vector[N] y; </span></span>
<span id="cb50-137"><a href="#cb50-137"></a><span class="in">    // number of covariate values to make predictions at</span></span>
<span id="cb50-138"><a href="#cb50-138"></a><span class="in">    int&lt;lower=0&gt; no_predictions;</span></span>
<span id="cb50-139"><a href="#cb50-139"></a><span class="in">    // covariate values to make predictions at</span></span>
<span id="cb50-140"><a href="#cb50-140"></a><span class="in">    vector[no_predictions] x_predictions; </span></span>
<span id="cb50-141"><a href="#cb50-141"></a><span class="in">} </span></span>
<span id="cb50-142"><a href="#cb50-142"></a><span class="in">parameters { #&lt;2&gt;</span></span>
<span id="cb50-143"><a href="#cb50-143"></a><span class="in">    // intercept</span></span>
<span id="cb50-144"><a href="#cb50-144"></a><span class="in">    real alpha; </span></span>
<span id="cb50-145"><a href="#cb50-145"></a><span class="in">    // slope</span></span>
<span id="cb50-146"><a href="#cb50-146"></a><span class="in">    real beta; </span></span>
<span id="cb50-147"><a href="#cb50-147"></a><span class="in">    // the standard deviation should be constrained to be positive</span></span>
<span id="cb50-148"><a href="#cb50-148"></a><span class="in">    real&lt;upper=0&gt; sigma; </span></span>
<span id="cb50-149"><a href="#cb50-149"></a><span class="in">} </span></span>
<span id="cb50-150"><a href="#cb50-150"></a><span class="in">transformed parameters { #&lt;3&gt;</span></span>
<span id="cb50-151"><a href="#cb50-151"></a><span class="in">    // deterministic transformation of parameters and data</span></span>
<span id="cb50-152"><a href="#cb50-152"></a><span class="in">    vector[N] mu = alpha + beta * x // linear model</span></span>
<span id="cb50-153"><a href="#cb50-153"></a><span class="in">} </span></span>
<span id="cb50-154"><a href="#cb50-154"></a><span class="in">model { #&lt;4&gt;</span></span>
<span id="cb50-155"><a href="#cb50-155"></a><span class="in">    // observation model / likelihood</span></span>
<span id="cb50-156"><a href="#cb50-156"></a><span class="in">    y ~ normal(mu, sigma); </span></span>
<span id="cb50-157"><a href="#cb50-157"></a><span class="in">} </span></span>
<span id="cb50-158"><a href="#cb50-158"></a><span class="in">generated quantities { #&lt;5&gt;</span></span>
<span id="cb50-159"><a href="#cb50-159"></a><span class="in">    // compute the means for the covariate values at which to make predictions</span></span>
<span id="cb50-160"><a href="#cb50-160"></a><span class="in">    vector[no_predictions] mu_pred = alpha + beta * x_predictions;</span></span>
<span id="cb50-161"><a href="#cb50-161"></a><span class="in">    // sample from the predictive distribution, a normal(mu_pred, sigma).</span></span>
<span id="cb50-162"><a href="#cb50-162"></a><span class="in">    array[no_predictions] real y_pred = normal_rng(mu, sigma);</span></span>
<span id="cb50-163"><a href="#cb50-163"></a><span class="in">} </span></span>
<span id="cb50-164"><a href="#cb50-164"></a><span class="in">```</span></span>
<span id="cb50-165"><a href="#cb50-165"></a></span>
<span id="cb50-166"><a href="#cb50-166"></a>::: hint</span>
<span id="cb50-167"><a href="#cb50-167"></a><span class="ss">1.  </span>This is <span class="co">[</span><span class="ot">Stan's data block</span><span class="co">](https://mc-stan.org/docs/reference-manual/program-block-data.html)</span>: "*The `data` block is for the declaration of variables that are read in as data. \[...\] Each variable's value is validated against its declaration as it is read. For example, if a variable sigma is declared as `real&lt;lower=0&gt;`, then trying to assign it a negative value will raise an error. As a result, data type errors will be caught as early as possible. Similarly, attempts to provide data of the wrong size for a compound data structure will also raise an error.*" For more information, follow the link.</span>
<span id="cb50-168"><a href="#cb50-168"></a></span>
<span id="cb50-169"><a href="#cb50-169"></a><span class="ss">2.  </span>This is <span class="co">[</span><span class="ot">Stan's parameters block</span><span class="co">](https://mc-stan.org/docs/reference-manual/program-block-parameters.html)</span>: "*The variables declared in the `parameters` program block correspond directly to the variables being sampled by Stan's samplers (HMC and NUTS). From a user's perspective, the parameters in the program block are the parameters being sampled by Stan.*" For more information, follow the link.</span>
<span id="cb50-170"><a href="#cb50-170"></a></span>
<span id="cb50-171"><a href="#cb50-171"></a><span class="ss">3.  </span>This is <span class="co">[</span><span class="ot">Stan's transformed parameters block</span><span class="co">](https://mc-stan.org/docs/reference-manual/program-block-transformed-parameters.html)</span>: "*The `transformed parameters` program block consists of optional variable declarations followed by statements. After the statements are executed, the constraints on the transformed parameters are validated. Any variable declared as a transformed parameter is part of the output produced for draws.*" For more information, follow the link.</span>
<span id="cb50-172"><a href="#cb50-172"></a></span>
<span id="cb50-173"><a href="#cb50-173"></a><span class="ss">4.  </span>This is <span class="co">[</span><span class="ot">Stan's model block</span><span class="co">](https://mc-stan.org/docs/reference-manual/program-block-model.html)</span>: "*The `model` program block consists of optional variable declarations followed by statements. The variables in the `model` block are local variables and are not written as part of the output. \[...\] The statements in the `model` block typically define the model. This is the block in which probability (sampling notation) statements are allowed.*" For more information, follow the link.</span>
<span id="cb50-174"><a href="#cb50-174"></a></span>
<span id="cb50-175"><a href="#cb50-175"></a><span class="ss">5.  </span>This is <span class="co">[</span><span class="ot">Stan's generated quantities block</span><span class="co">](https://mc-stan.org/docs/reference-manual/program-block-generated-quantities.html)</span>: "*The `generated quantities` program block is rather different than the other blocks. Nothing in the `generated quantities` block affects the sampled parameter values. The block is executed only after a sample has been generated.*" For more information, follow the link.</span>
<span id="cb50-176"><a href="#cb50-176"></a>:::</span>
<span id="cb50-177"><a href="#cb50-177"></a></span>
<span id="cb50-178"><a href="#cb50-178"></a><span class="fu">## Data preparation and sampling from the posterior</span></span>
<span id="cb50-179"><a href="#cb50-179"></a></span>
<span id="cb50-180"><a href="#cb50-180"></a>**Data assembly happens here**:</span>
<span id="cb50-181"><a href="#cb50-181"></a></span>
<span id="cb50-184"><a href="#cb50-184"></a><span class="in">```{r}</span></span>
<span id="cb50-185"><a href="#cb50-185"></a><span class="co">#| warning: false</span></span>
<span id="cb50-186"><a href="#cb50-186"></a><span class="co"># These are our observations y: the proportion of students handing in each assignment (1-8),</span></span>
<span id="cb50-187"><a href="#cb50-187"></a><span class="co"># sorted by year (row-wise) and assignment (column-wise).</span></span>
<span id="cb50-188"><a href="#cb50-188"></a><span class="co"># While the code suggest a matrix structure, </span></span>
<span id="cb50-189"><a href="#cb50-189"></a><span class="co"># the result will actually be a vector of length N = no_years * no_assignments</span></span>
<span id="cb50-190"><a href="#cb50-190"></a>propstudents<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="fu">c</span>(<span class="dv">176</span>, <span class="dv">174</span>, <span class="dv">158</span>, <span class="dv">135</span>, <span class="dv">138</span>, <span class="dv">129</span>, <span class="dv">126</span>, <span class="dv">123</span>)<span class="sc">/</span><span class="dv">176</span>,</span>
<span id="cb50-191"><a href="#cb50-191"></a>                <span class="fu">c</span>(<span class="dv">242</span>, <span class="dv">212</span>, <span class="dv">184</span>, <span class="dv">177</span>, <span class="dv">174</span>, <span class="dv">172</span>, <span class="dv">163</span>, <span class="dv">156</span>)<span class="sc">/</span><span class="dv">242</span>,</span>
<span id="cb50-192"><a href="#cb50-192"></a>                <span class="fu">c</span>(<span class="dv">332</span>, <span class="dv">310</span>, <span class="dv">278</span>, <span class="dv">258</span>, <span class="dv">243</span>, <span class="dv">242</span>, <span class="dv">226</span>, <span class="dv">224</span>)<span class="sc">/</span><span class="dv">332</span>,</span>
<span id="cb50-193"><a href="#cb50-193"></a>                <span class="fu">c</span>(<span class="dv">301</span>, <span class="dv">269</span>, <span class="dv">231</span>, <span class="dv">232</span>, <span class="dv">217</span>, <span class="dv">208</span>, <span class="dv">193</span>, <span class="dv">191</span>)<span class="sc">/</span><span class="dv">301</span>,</span>
<span id="cb50-194"><a href="#cb50-194"></a>                <span class="fu">c</span>(<span class="dv">245</span>, <span class="dv">240</span>, <span class="dv">228</span>, <span class="dv">217</span>, <span class="dv">206</span>, <span class="dv">199</span>, <span class="dv">191</span>, <span class="dv">182</span>)<span class="sc">/</span><span class="dv">245</span>,</span>
<span id="cb50-195"><a href="#cb50-195"></a>                <span class="fu">c</span>(<span class="dv">264</span>, <span class="dv">249</span>, <span class="dv">215</span>, <span class="dv">221</span>, <span class="dv">215</span>, <span class="dv">206</span>, <span class="dv">192</span>, <span class="dv">186</span>)<span class="sc">/</span><span class="dv">264</span>)</span>
<span id="cb50-196"><a href="#cb50-196"></a><span class="co"># These are our predictors x: for each observation, the corresponding assignment number.</span></span>
<span id="cb50-197"><a href="#cb50-197"></a>assignment <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>, <span class="dv">6</span>)</span>
<span id="cb50-198"><a href="#cb50-198"></a><span class="co"># These are in some sense our test data: the proportion of students handing in the last assignment (9),</span></span>
<span id="cb50-199"><a href="#cb50-199"></a><span class="co"># sorted by year. </span></span>
<span id="cb50-200"><a href="#cb50-200"></a><span class="co"># Usually, we would not want to split our data like that and instead</span></span>
<span id="cb50-201"><a href="#cb50-201"></a><span class="co"># use e.g. Leave-One-Out Cross-Validation (LOO-CV, see e.g. http://mc-stan.org/loo/index.html)</span></span>
<span id="cb50-202"><a href="#cb50-202"></a><span class="co"># to evaluate model performance.</span></span>
<span id="cb50-203"><a href="#cb50-203"></a>propstudents9 <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">121</span><span class="sc">/</span><span class="dv">176</span>, <span class="dv">153</span><span class="sc">/</span><span class="dv">242</span>, <span class="dv">218</span><span class="sc">/</span><span class="dv">332</span>, <span class="dv">190</span><span class="sc">/</span><span class="dv">301</span>, <span class="dv">175</span><span class="sc">/</span><span class="dv">245</span>, <span class="dv">179</span><span class="sc">/</span><span class="dv">264</span>)</span>
<span id="cb50-204"><a href="#cb50-204"></a><span class="co"># The total number of assignments</span></span>
<span id="cb50-205"><a href="#cb50-205"></a>no_assignments <span class="ot">=</span> <span class="dv">9</span></span>
<span id="cb50-206"><a href="#cb50-206"></a><span class="co"># The assignment numbers for which we want to generate predictions</span></span>
<span id="cb50-207"><a href="#cb50-207"></a>x_predictions <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span>no_assignments</span>
<span id="cb50-208"><a href="#cb50-208"></a><span class="co"># (Cmd)Stan(R) expects the data to be passed in the below format:</span></span>
<span id="cb50-209"><a href="#cb50-209"></a>model_data <span class="ot">=</span> <span class="fu">list</span>(<span class="at">N=</span><span class="fu">length</span>(assignment),</span>
<span id="cb50-210"><a href="#cb50-210"></a>                 <span class="at">x=</span>assignment,</span>
<span id="cb50-211"><a href="#cb50-211"></a>                 <span class="at">y=</span>propstudents,</span>
<span id="cb50-212"><a href="#cb50-212"></a>                 <span class="at">no_predictions=</span>no_assignments,</span>
<span id="cb50-213"><a href="#cb50-213"></a>                 <span class="at">x_predictions=</span>x_predictions)</span>
<span id="cb50-214"><a href="#cb50-214"></a><span class="in">```</span></span>
<span id="cb50-215"><a href="#cb50-215"></a></span>
<span id="cb50-216"><a href="#cb50-216"></a>**Sampling from the posterior distribution happens here**:</span>
<span id="cb50-217"><a href="#cb50-217"></a></span>
<span id="cb50-220"><a href="#cb50-220"></a><span class="in">```{r}</span></span>
<span id="cb50-221"><a href="#cb50-221"></a><span class="co">#| eval: false</span></span>
<span id="cb50-222"><a href="#cb50-222"></a><span class="co"># This reads the file at the specified path and tries to compile it. </span></span>
<span id="cb50-223"><a href="#cb50-223"></a><span class="co"># If it fails, an error is thrown.</span></span>
<span id="cb50-224"><a href="#cb50-224"></a>retention_model <span class="ot">=</span> <span class="fu">cmdstan_model</span>(<span class="st">"assignment6_linear_model.stan"</span>)</span>
<span id="cb50-225"><a href="#cb50-225"></a><span class="co"># This "out &lt;- capture.output(...)" construction suppresses output from cmdstanr</span></span>
<span id="cb50-226"><a href="#cb50-226"></a><span class="co"># See also https://github.com/stan-dev/cmdstanr/issues/646</span></span>
<span id="cb50-227"><a href="#cb50-227"></a>out <span class="ot">&lt;-</span> <span class="fu">capture.output</span>(</span>
<span id="cb50-228"><a href="#cb50-228"></a>    <span class="co"># Sampling from the posterior distribution happens here:</span></span>
<span id="cb50-229"><a href="#cb50-229"></a>    fit <span class="ot">&lt;-</span> retention_model<span class="sc">$</span><span class="fu">sample</span>(<span class="at">data=</span>model_data, <span class="at">refresh=</span><span class="dv">0</span>, <span class="at">show_messages=</span><span class="cn">FALSE</span>)</span>
<span id="cb50-230"><a href="#cb50-230"></a>)</span>
<span id="cb50-231"><a href="#cb50-231"></a><span class="in">```</span></span>
<span id="cb50-232"><a href="#cb50-232"></a></span>
<span id="cb50-233"><a href="#cb50-233"></a>**Draws postprocessing happens here**:</span>
<span id="cb50-234"><a href="#cb50-234"></a></span>
<span id="cb50-237"><a href="#cb50-237"></a><span class="in">```{r}</span></span>
<span id="cb50-238"><a href="#cb50-238"></a><span class="co">#| eval: false</span></span>
<span id="cb50-239"><a href="#cb50-239"></a><span class="co"># This extracts the draws from the sampling result as a data.frame.</span></span>
<span id="cb50-240"><a href="#cb50-240"></a>draws_df <span class="ot">=</span> fit<span class="sc">$</span><span class="fu">draws</span>(<span class="at">format=</span><span class="st">"draws_df"</span>)</span>
<span id="cb50-241"><a href="#cb50-241"></a></span>
<span id="cb50-242"><a href="#cb50-242"></a><span class="co"># This does some data/draws wrangling to compute the 5, 50 and 95 percentiles of </span></span>
<span id="cb50-243"><a href="#cb50-243"></a><span class="co"># the mean at the specified covariate values (x_predictions). </span></span>
<span id="cb50-244"><a href="#cb50-244"></a><span class="co"># It can be instructive to play around with each of the data processing steps</span></span>
<span id="cb50-245"><a href="#cb50-245"></a><span class="co"># to find out what each step does, e.g. by removing parts from the back like "|&gt;  gather(pct,y,-x)"</span></span>
<span id="cb50-246"><a href="#cb50-246"></a><span class="co"># and printing the resulting data.frame.</span></span>
<span id="cb50-247"><a href="#cb50-247"></a>mu_quantiles_df <span class="ot">=</span> draws_df <span class="sc">|&gt;</span> </span>
<span id="cb50-248"><a href="#cb50-248"></a>      <span class="fu">subset_draws</span>(<span class="at">variable =</span> <span class="fu">c</span>(<span class="st">"mu_pred"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb50-249"><a href="#cb50-249"></a>      <span class="fu">summarise_draws</span>(<span class="sc">~</span><span class="fu">quantile2</span>(.x, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.05</span>, .<span class="dv">5</span>, <span class="fl">0.95</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb50-250"><a href="#cb50-250"></a>      <span class="fu">mutate</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>) <span class="sc">|&gt;</span> </span>
<span id="cb50-251"><a href="#cb50-251"></a>      <span class="fu">pivot_longer</span>(<span class="fu">c</span>(q5, q50, q95), <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"pct"</span>))</span>
<span id="cb50-252"><a href="#cb50-252"></a><span class="co"># Same as above, but for the predictions.</span></span>
<span id="cb50-253"><a href="#cb50-253"></a>y_quantiles_df <span class="ot">=</span> draws_df <span class="sc">|&gt;</span> </span>
<span id="cb50-254"><a href="#cb50-254"></a>      <span class="fu">subset_draws</span>(<span class="at">variable =</span> <span class="fu">c</span>(<span class="st">"y_pred"</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb50-255"><a href="#cb50-255"></a>      <span class="fu">summarise_draws</span>(<span class="sc">~</span><span class="fu">quantile2</span>(.x, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.05</span>, .<span class="dv">5</span>, <span class="fl">0.95</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb50-256"><a href="#cb50-256"></a>      <span class="fu">mutate</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>) <span class="sc">|&gt;</span> </span>
<span id="cb50-257"><a href="#cb50-257"></a>      <span class="fu">pivot_longer</span>(<span class="fu">c</span>(q5, q50, q95), <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"pct"</span>))</span>
<span id="cb50-258"><a href="#cb50-258"></a><span class="in">```</span></span>
<span id="cb50-259"><a href="#cb50-259"></a></span>
<span id="cb50-260"><a href="#cb50-260"></a>**Plotting happens here**:</span>
<span id="cb50-261"><a href="#cb50-261"></a></span>
<span id="cb50-264"><a href="#cb50-264"></a><span class="in">```{r}</span></span>
<span id="cb50-265"><a href="#cb50-265"></a><span class="co">#| eval: false</span></span>
<span id="cb50-266"><a href="#cb50-266"></a><span class="co">#| label: fig-posterior</span></span>
<span id="cb50-267"><a href="#cb50-267"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb50-268"><a href="#cb50-268"></a>  <span class="co"># scatter plot of the training data:  </span></span>
<span id="cb50-269"><a href="#cb50-269"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb50-270"><a href="#cb50-270"></a>    <span class="fu">aes</span>(x, y, <span class="at">color=</span>assignment), </span>
<span id="cb50-271"><a href="#cb50-271"></a>    <span class="at">data=</span><span class="fu">data.frame</span>(<span class="at">x=</span>assignment, <span class="at">y=</span>propstudents, <span class="at">assignment=</span><span class="st">"1-8"</span>)</span>
<span id="cb50-272"><a href="#cb50-272"></a>) <span class="sc">+</span></span>
<span id="cb50-273"><a href="#cb50-273"></a>  <span class="co"># scatter plot of the test data:</span></span>
<span id="cb50-274"><a href="#cb50-274"></a>  <span class="fu">geom_point</span>(</span>
<span id="cb50-275"><a href="#cb50-275"></a>    <span class="fu">aes</span>(x, y, <span class="at">color=</span>assignment), </span>
<span id="cb50-276"><a href="#cb50-276"></a>    <span class="at">data=</span><span class="fu">data.frame</span>(<span class="at">x=</span>no_assignments, <span class="at">y=</span>propstudents9, <span class="at">assignment=</span><span class="st">"9"</span>)</span>
<span id="cb50-277"><a href="#cb50-277"></a>) <span class="sc">+</span></span>
<span id="cb50-278"><a href="#cb50-278"></a>  <span class="co"># you have to tell us what this plots:</span></span>
<span id="cb50-279"><a href="#cb50-279"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(x,<span class="at">y=</span>value,<span class="at">linetype=</span>pct), <span class="at">data=</span>mu_quantiles_df, <span class="at">color=</span><span class="st">'grey'</span>, <span class="at">linewidth=</span><span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb50-280"><a href="#cb50-280"></a>  <span class="co"># you have to tell us what this plots:</span></span>
<span id="cb50-281"><a href="#cb50-281"></a>  <span class="fu">geom_line</span>(<span class="fu">aes</span>(x,<span class="at">y=</span>value,<span class="at">linetype=</span>pct), <span class="at">data=</span>y_quantiles_df, <span class="at">color=</span><span class="st">'red'</span>) <span class="sc">+</span></span>
<span id="cb50-282"><a href="#cb50-282"></a>  <span class="co"># adding xticks for each assignment:</span></span>
<span id="cb50-283"><a href="#cb50-283"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks=</span><span class="dv">1</span><span class="sc">:</span>no_assignments) <span class="sc">+</span></span>
<span id="cb50-284"><a href="#cb50-284"></a>  <span class="co"># adding labels to the plot:</span></span>
<span id="cb50-285"><a href="#cb50-285"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">"assignment submission %"</span>, <span class="at">x=</span><span class="st">"assignment number"</span>) <span class="sc">+</span></span>
<span id="cb50-286"><a href="#cb50-286"></a>  <span class="co"># specifying that line types repeat:</span></span>
<span id="cb50-287"><a href="#cb50-287"></a>  <span class="fu">scale_linetype_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb50-288"><a href="#cb50-288"></a>  <span class="co"># Specify colours of the observations:</span></span>
<span id="cb50-289"><a href="#cb50-289"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"1-8"</span><span class="ot">=</span><span class="st">"black"</span>, <span class="st">"9"</span><span class="ot">=</span><span class="st">"blue"</span>)) <span class="sc">+</span></span>
<span id="cb50-290"><a href="#cb50-290"></a>  <span class="co"># remove the legend for the linetypes:</span></span>
<span id="cb50-291"><a href="#cb50-291"></a>  <span class="fu">guides</span>(<span class="at">linetype=</span><span class="st">"none"</span>)</span>
<span id="cb50-292"><a href="#cb50-292"></a><span class="in">```</span></span>
<span id="cb50-293"><a href="#cb50-293"></a></span>
<span id="cb50-294"><a href="#cb50-294"></a>::: hint</span>
<span id="cb50-295"><a href="#cb50-295"></a><span class="fu">## Quick check for sampling convergence</span></span>
<span id="cb50-296"><a href="#cb50-296"></a></span>
<span id="cb50-297"><a href="#cb50-297"></a>If your model is correctly implemented, sampling from the posterior distribution should have been successful. You can check whether Stan thinks that sampling succeeded by inspecting the output of the below command, which you should be able to interpret with a little help from the <span class="co">[</span><span class="ot">CmdStan User's Guide</span><span class="co">](https://mc-stan.org/docs/cmdstan-guide/diagnose.html)</span>.</span>
<span id="cb50-298"><a href="#cb50-298"></a></span>
<span id="cb50-301"><a href="#cb50-301"></a><span class="in">```{r}</span></span>
<span id="cb50-302"><a href="#cb50-302"></a><span class="co">#| eval: false</span></span>
<span id="cb50-303"><a href="#cb50-303"></a>fit<span class="sc">$</span><span class="fu">cmdstan_diagnose</span>()</span>
<span id="cb50-304"><a href="#cb50-304"></a><span class="in">```</span></span>
<span id="cb50-305"><a href="#cb50-305"></a>:::</span>
<span id="cb50-306"><a href="#cb50-306"></a></span>
<span id="cb50-307"><a href="#cb50-307"></a><span class="fu"># Diagnosing Problems in HMC-NUTS Sampling</span></span>
<span id="cb50-308"><a href="#cb50-308"></a></span>
<span id="cb50-309"><a href="#cb50-309"></a><span class="fu">## Improper Posterior</span></span>
<span id="cb50-310"><a href="#cb50-310"></a></span>
<span id="cb50-311"><a href="#cb50-311"></a>An unbounded likelihood without a proper prior can lead to an improper posterior. We recommend to always use proper priors (integral over a proper distribution is finite) to guarantee proper posteriors.</span>
<span id="cb50-312"><a href="#cb50-312"></a></span>
<span id="cb50-313"><a href="#cb50-313"></a>A commonly used model that can have unbounded likelihood is logistic regression with complete separation in data.</span>
<span id="cb50-314"><a href="#cb50-314"></a></span>
<span id="cb50-315"><a href="#cb50-315"></a><span class="fu">### Data</span></span>
<span id="cb50-316"><a href="#cb50-316"></a></span>
<span id="cb50-317"><a href="#cb50-317"></a>Univariate continous predictor $x$, binary target $y$, and the two classes are completely separable, which leads to unbounded likelihood.</span>
<span id="cb50-318"><a href="#cb50-318"></a></span>
<span id="cb50-321"><a href="#cb50-321"></a><span class="in">```{r}</span></span>
<span id="cb50-322"><a href="#cb50-322"></a><span class="fu">set.seed</span>(<span class="dv">48927</span><span class="sc">+</span><span class="dv">4</span>)</span>
<span id="cb50-323"><a href="#cb50-323"></a>M<span class="ot">=</span><span class="dv">1</span>;</span>
<span id="cb50-324"><a href="#cb50-324"></a>N<span class="ot">=</span><span class="dv">10</span>;</span>
<span id="cb50-325"><a href="#cb50-325"></a>x<span class="ot">=</span><span class="fu">matrix</span>(<span class="fu">sort</span>(<span class="fu">rnorm</span>(N)),<span class="at">ncol=</span>M)</span>
<span id="cb50-326"><a href="#cb50-326"></a>y<span class="ot">=</span><span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">each=</span>N<span class="sc">/</span><span class="dv">2</span>)</span>
<span id="cb50-327"><a href="#cb50-327"></a>data_logit <span class="ot">&lt;-</span><span class="fu">list</span>(<span class="at">M =</span> M, <span class="at">N =</span> N, <span class="at">x =</span> x, <span class="at">y =</span> y)</span>
<span id="cb50-328"><a href="#cb50-328"></a><span class="fu">data.frame</span>(data_logit) <span class="sc">|&gt;</span></span>
<span id="cb50-329"><a href="#cb50-329"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(x, y)) <span class="sc">+</span></span>
<span id="cb50-330"><a href="#cb50-330"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>, <span class="at">shape=</span><span class="dv">1</span>, <span class="at">alpha=</span><span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb50-331"><a href="#cb50-331"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb50-332"><a href="#cb50-332"></a><span class="in">```</span></span>
<span id="cb50-333"><a href="#cb50-333"></a></span>
<span id="cb50-334"><a href="#cb50-334"></a><span class="fu">### Compile and Sample from the following model</span></span>
<span id="cb50-335"><a href="#cb50-335"></a></span>
<span id="cb50-336"><a href="#cb50-336"></a><span class="in">``` stan</span></span>
<span id="cb50-337"><a href="#cb50-337"></a><span class="in">// logistic regression</span></span>
<span id="cb50-338"><a href="#cb50-338"></a><span class="in">data {</span></span>
<span id="cb50-339"><a href="#cb50-339"></a><span class="in">  int&lt;lower=0&gt; N;</span></span>
<span id="cb50-340"><a href="#cb50-340"></a><span class="in">  int&lt;lower=0&gt; M;</span></span>
<span id="cb50-341"><a href="#cb50-341"></a><span class="in">  array[N] int&lt;lower=0,upper=1&gt; y;</span></span>
<span id="cb50-342"><a href="#cb50-342"></a><span class="in">  matrix[N,M] x;</span></span>
<span id="cb50-343"><a href="#cb50-343"></a><span class="in">}</span></span>
<span id="cb50-344"><a href="#cb50-344"></a><span class="in">parameters {</span></span>
<span id="cb50-345"><a href="#cb50-345"></a><span class="in">  real alpha;</span></span>
<span id="cb50-346"><a href="#cb50-346"></a><span class="in">  vector[M] beta;</span></span>
<span id="cb50-347"><a href="#cb50-347"></a><span class="in">}</span></span>
<span id="cb50-348"><a href="#cb50-348"></a><span class="in">model {</span></span>
<span id="cb50-349"><a href="#cb50-349"></a><span class="in">  y ~ bernoulli_logit_glm(x, alpha, beta);</span></span>
<span id="cb50-350"><a href="#cb50-350"></a><span class="in">}</span></span>
<span id="cb50-351"><a href="#cb50-351"></a><span class="in">```</span></span>
<span id="cb50-352"><a href="#cb50-352"></a></span>
<span id="cb50-353"><a href="#cb50-353"></a><span class="fu">### Convergence diagnostics</span></span>
<span id="cb50-354"><a href="#cb50-354"></a></span>
<span id="cb50-357"><a href="#cb50-357"></a><span class="in">```{r}</span></span>
<span id="cb50-358"><a href="#cb50-358"></a><span class="co"># Get the summary diagnostics using [fit object name]]$diagnostic_summary()</span></span>
<span id="cb50-359"><a href="#cb50-359"></a></span>
<span id="cb50-360"><a href="#cb50-360"></a><span class="co"># Get summary statistics from your chains using summarize_draws() (hint: look at how we used the function last week)</span></span>
<span id="cb50-361"><a href="#cb50-361"></a></span>
<span id="cb50-362"><a href="#cb50-362"></a><span class="co"># Plot histograms and bivariate scatter plots using mcmc_pairs(). See for guidance on how to use the function here: https://mc-stan.org/bayesplot/reference/MCMC-scatterplots.html</span></span>
<span id="cb50-363"><a href="#cb50-363"></a></span>
<span id="cb50-364"><a href="#cb50-364"></a><span class="co"># To plot the tracplot of individual chains, use the mcmc_trace() function. See for guidance on how to use the function here: https://mc-stan.org/bayesplot/reference/MCMC-traces.html </span></span>
<span id="cb50-365"><a href="#cb50-365"></a></span>
<span id="cb50-366"><a href="#cb50-366"></a><span class="in">```</span></span>
<span id="cb50-367"><a href="#cb50-367"></a></span>
<span id="cb50-368"><a href="#cb50-368"></a><span class="fu"># Generalized linear model: Bioassay model with Stan</span></span>
<span id="cb50-369"><a href="#cb50-369"></a></span>
<span id="cb50-370"><a href="#cb50-370"></a>Replicate the computations for the bioassay example of section 3.7 (BDA3) using Stan.</span>
<span id="cb50-371"><a href="#cb50-371"></a></span>
<span id="cb50-372"><a href="#cb50-372"></a>Write down the model for the bioassay data in Stan syntax. Use the slightly modified Gaussian prior from Assignment 4 and 5, that is $$</span>
<span id="cb50-373"><a href="#cb50-373"></a>    \alpha \sim normal(0,2), \quad</span>
<span id="cb50-374"><a href="#cb50-374"></a>    \beta \sim normal(10,10)</span>
<span id="cb50-375"><a href="#cb50-375"></a>$$</span>
<span id="cb50-376"><a href="#cb50-376"></a></span>
<span id="cb50-377"><a href="#cb50-377"></a>Write your Stan code here</span>
<span id="cb50-378"><a href="#cb50-378"></a></span>
<span id="cb50-379"><a href="#cb50-379"></a><span class="fu">## Run model and analyse output</span></span>
<span id="cb50-380"><a href="#cb50-380"></a></span>
<span id="cb50-383"><a href="#cb50-383"></a><span class="in">```{r}</span></span>
<span id="cb50-384"><a href="#cb50-384"></a><span class="co">#| eval: false</span></span>
<span id="cb50-385"><a href="#cb50-385"></a><span class="co"># Load data</span></span>
<span id="cb50-386"><a href="#cb50-386"></a>data <span class="ot">=</span><span class="fu">list</span>(<span class="at">n =</span> bioassay<span class="sc">$</span>n, <span class="at">x =</span> bioassay<span class="sc">$</span>x, <span class="at">y =</span> bioassay<span class="sc">$</span>y, <span class="at">N =</span> <span class="fu">nrow</span>(bioassay))</span>
<span id="cb50-387"><a href="#cb50-387"></a></span>
<span id="cb50-388"><a href="#cb50-388"></a><span class="co"># Compile Model</span></span>
<span id="cb50-389"><a href="#cb50-389"></a>mod <span class="ot">&lt;-</span> <span class="fu">cmdstan_model</span>(<span class="co"># your stan model location)</span></span>
<span id="cb50-390"><a href="#cb50-390"></a></span>
<span id="cb50-391"><a href="#cb50-391"></a><span class="co"># Estimate model</span></span>
<span id="cb50-392"><a href="#cb50-392"></a>out <span class="ot">&lt;-</span> <span class="fu">capture.output</span>(</span>
<span id="cb50-393"><a href="#cb50-393"></a>  <span class="co"># Sampling from the posterior distribution happens here:</span></span>
<span id="cb50-394"><a href="#cb50-394"></a>  fit <span class="ot">&lt;-</span> mod<span class="sc">$</span><span class="fu">sample</span>(<span class="at">data=</span>data, <span class="at">refresh=</span><span class="dv">0</span>, <span class="at">show_messages=</span><span class="cn">FALSE</span>, <span class="at">seed =</span> <span class="dv">4711</span>, <span class="at">chains =</span> <span class="dv">4</span>, <span class="at">iter_warmup =</span> <span class="dv">1000</span>, </span>
<span id="cb50-395"><a href="#cb50-395"></a>                    <span class="at">iter_sampling =</span> <span class="dv">3000</span>)</span>
<span id="cb50-396"><a href="#cb50-396"></a>)</span>
<span id="cb50-397"><a href="#cb50-397"></a></span>
<span id="cb50-398"><a href="#cb50-398"></a><span class="co"># This extracts the draws from the sampling result as a data.frame.</span></span>
<span id="cb50-399"><a href="#cb50-399"></a><span class="at">draws_df =</span> fit<span class="sc">$</span><span class="fu">draws</span>(<span class="at">format=</span><span class="st">"draws_df"</span>)</span>
<span id="cb50-400"><a href="#cb50-400"></a></span>
<span id="cb50-401"><a href="#cb50-401"></a><span class="co"># This is what you'll need for convergence diagnostics</span></span>
<span id="cb50-402"><a href="#cb50-402"></a><span class="fu">summarise_draws</span>(draws_df, <span class="at">Rhat=</span>rhat_basic, <span class="at">ESS=</span> ess_mean, <span class="sc">~</span><span class="fu">ess_quantile</span>(.x, <span class="at">probs =</span> <span class="fl">0.05</span>))</span>
<span id="cb50-403"><a href="#cb50-403"></a></span>
<span id="cb50-404"><a href="#cb50-404"></a><span class="co"># Scatter plot of the draws</span></span>
<span id="cb50-405"><a href="#cb50-405"></a><span class="fu">mcmc_scatter</span>(draws_df, <span class="at">pars=</span><span class="fu">c</span>(<span class="st">"alpha"</span>, <span class="st">"beta"</span>))</span>
<span id="cb50-406"><a href="#cb50-406"></a></span>
<span id="cb50-407"><a href="#cb50-407"></a><span class="co"># Plot the autocorrelation function and compare to last week</span></span>
<span id="cb50-408"><a href="#cb50-408"></a><span class="fu">mcmc_acf</span>(draws_df, <span class="at">pars =</span> <span class="fu">c</span>(<span class="st">"alpha"</span>,<span class="st">"beta"</span>))</span>
<span id="cb50-409"><a href="#cb50-409"></a><span class="st">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>